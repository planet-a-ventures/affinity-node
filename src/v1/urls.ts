/**
 * @hidden
 * See [here](https://api-docs.affinity.co/#get-a-specific-list) for more info.
 */
export const listsUrl = (list_id?: number) =>
    list_id ? `/lists/${encodeURIComponent(list_id)}` : '/lists'

/**
 * @hidden
 * See [here](https://api-docs.affinity.co/#get-a-specific-list-entry) for more info.
 */
export const listEntriesUrl = (list_id: number, list_entry_id?: number) =>
    listsUrl(list_id) + (
        list_entry_id
            ? `/list-entries/${encodeURIComponent(list_entry_id)}`
            : '/list-entries'
    )

/**
 * @hidden
 * See [here](https://api-docs.affinity.co/#rate-limit) for more info.
 */
export const rateLimitUrl = () => '/rate-limit'

/**
 * @hidden
 * See [here](https://api-docs.affinity.co/#whoami) for more info.
 */
export const whoAmIUrl = () => '/auth/whoami'

/**
 * @hidden
 * See [here](https://api-docs.affinity.co/#fields) for more info.
 */
export const fieldsUrl = (field_id?: number) =>
    field_id ? `/fields/${encodeURIComponent(field_id)}` : '/fields'

/**
 * @hidden
 * See [here](https://api-docs.affinity.co/#field-values) for more info.
 */
export const fieldValuesUrl = (field_value_id?: number) => {
    return field_value_id
        ? `/field-values/${encodeURIComponent(field_value_id)}`
        : '/field-values'
}

/**
 * @hidden
 * See [here](https://api-docs.affinity.co/#get-field-values-changes) for more info.
 */
export const fieldValueChangesUrl = () => '/field-value-changes'

/**
 * @hidden
 * See [here](https://api-docs.affinity.co/#organizations) for more info.
 */
export const organizationsUrl = (organization_id?: number | 'fields') => {
    return organization_id
        ? `/organizations/${encodeURIComponent(organization_id)}`
        : '/organizations'
}

/**
 * @hidden
 * See [here](https://api-docs.affinity.co/#get-global-organizations-fields) for more info.
 */
export const organizationFieldsUrl = () => organizationsUrl('fields')

/**
 * @hidden
 * See [here](https://api-docs.affinity.co/#persons) for more info.
 */
export const personsUrl = (person_id?: number | 'fields') => {
    return person_id ? `/persons/${encodeURIComponent(person_id)}` : '/persons'
}

/**
 * @hidden
 * See [here](https://api-docs.affinity.co/#get-global-person-fields) for more info.
 */
export const personFieldsUrl = () => personsUrl('fields')

/**
 * @hidden
 * See [here](https://api-docs.affinity.co/#entity-files) for more info.
 */
export const entityFilesUrl = (
    entity_file_id?: number,
    is_download: boolean = false,
) => {
    return entity_file_id
        ? `/entity-files` + (is_download ? '/download' : '') +
            `/${encodeURIComponent(entity_file_id)}`
        : '/entity-files'
}
