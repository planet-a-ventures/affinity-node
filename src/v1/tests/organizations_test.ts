import { assertSnapshot } from '@std/testing/snapshot'
import { afterEach, beforeEach, describe, it } from '@std/testing/bdd'

import axios from 'axios'
import MockAdapter from 'axios-mock-adapter'
import { Affinity } from '../index.ts'
import { getRawFixture } from './get_raw_fixture.ts'
import { apiKey, isLiveRun } from './env.ts'
import { organizationFieldsUrl, organizationsUrl } from '../urls.ts'
import type { SearchOrganizationsRequest } from '../organizations.ts'
import { mockPagingFromAllEndpoint } from './mock_paging_from_all_endpoint.ts'

describe('organizations', () => {
    let mock: MockAdapter
    let affinity: Affinity

    beforeEach(() => {
        if (!isLiveRun()) {
            mock = new MockAdapter(axios, { onNoMatch: 'throwException' })
        }
        affinity = new Affinity(apiKey() || 'api_key')
    })
    afterEach(() => {
        mock?.reset()
    })

    it('can get a specific organization', async (t) => {
        const organization_id = 64779194
        mock?.onGet(organizationsUrl(organization_id)).reply(
            200,
            await getRawFixture('organizations/get.raw.response.json'),
        )
        const res = await affinity.organizations.get({ organization_id })
        await assertSnapshot(t, res)
    })

    it('can search for organizations', async (t) => {
        const request = { term: 'affinity' }
        mock?.onGet(organizationsUrl(), { params: request }).reply(
            200,
            await getRawFixture('organizations/search.raw.response.json'),
        )
        const res = await affinity.organizations.search(request)
        await assertSnapshot(t, res)
    })

    it('can search for organizations with the appropriate dates', async (t) => {
        const myDate = new Date(1717428411010)
        const request: SearchOrganizationsRequest = {
            min_first_email_date: myDate,
            term: 'affinity',
            with_interaction_dates: true,
        }
        mock?.onGet(organizationsUrl(), {
            params: {
                term: request.term,
                min_first_email_date: myDate.toISOString(),
                with_interaction_dates: true,
            },
        }).reply(
            200,
            await getRawFixture('organizations/search.raw.response.json'),
        )
        const res = await affinity.organizations.search(request)
        await assertSnapshot(t, res)
    })

    it('can create a new organization', async (t) => {
        const data = {
            name: 'Acme Corporation',
            domain: 'acme.co',
            person_ids: [38706],
        }
        mock?.onPost(organizationsUrl()).reply(
            201,
            await getRawFixture('organizations/create.raw.response.json'),
        )
        const res = await affinity.organizations.create(data)
        await assertSnapshot(t, res)
    })

    it('can update an organization', async (t) => {
        const data = {
            organization_id: 120611418,
            name: 'Acme Corp.',
            person_ids: [38706, 89734],
        }
        mock?.onPut(organizationsUrl(data.organization_id)).reply(
            200,
            await getRawFixture('organizations/update.raw.response.json'),
        )
        const res = await affinity.organizations.update(data)
        await assertSnapshot(t, res)
    })

    it('can delete an organization', async (t) => {
        const organization_id = 120611418
        mock?.onDelete(organizationsUrl(organization_id)).reply(200, {
            success: true,
        })
        const res = await affinity.organizations.delete({ organization_id })
        await assertSnapshot(t, res)
    })

    it('can get global organization fields', async (t) => {
        mock?.onGet(organizationFieldsUrl()).reply(
            200,
            await getRawFixture('organizations/get_fields.raw.response.json'),
        )
        const res = await affinity.organizations.getFields()
        await assertSnapshot(t, res)
    })

    it('iterates over all organizations', async (t) => {
        const params: SearchOrganizationsRequest = {
            term: 'affinity',
            page_size: 1,
        }

        await mockPagingFromAllEndpoint(
            './fixtures/organizations/search.raw.response.json',
            params,
            'organizations',
            organizationsUrl,
            mock,
        )

        let page = 0
        for await (
            const entries of affinity.organizations.searchIterator(params)
        ) {
            await assertSnapshot(t, entries, {
                name: `page ${++page} of organizations`,
            })
        }
    })
})
