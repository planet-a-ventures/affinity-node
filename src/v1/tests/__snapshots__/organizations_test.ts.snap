export const snapshot = {};

snapshot[`organizations > can get a specific organization 1`] = `
{
  domain: "affinity.co",
  domains: [
    "affinity.co",
  ],
  global: false,
  id: 64779194,
  interaction_dates: {
    first_email_date: 2011-11-23T16:12:45.328Z,
    first_event_date: 2012-01-01T16:18:00.329Z,
    last_chat_message_date: 2011-11-01T10:26:56.537Z,
    last_email_date: 2012-03-04T13:06:07.890Z,
    last_event_date: 2011-12-11T10:26:56.537Z,
    last_interaction_date: 2012-03-04T13:06:07.890Z,
    next_event_date: 2019-03-04T13:06:07.890Z,
  },
  interactions: {
    first_email: {
      date: 2011-11-23T16:12:45.328Z,
      person_ids: [
        123,
      ],
    },
    first_event: {
      date: 2012-01-01T16:18:00.329Z,
      person_ids: [
        123,
      ],
    },
    last_chat_message: {
      date: 2011-11-01T10:26:56.537Z,
      person_ids: [
        123,
      ],
    },
    last_email: {
      date: 2012-03-04T13:06:07.890Z,
      person_ids: [
        123,
      ],
    },
    last_event: {
      date: 2011-12-11T10:26:56.537Z,
      person_ids: [
        123,
      ],
    },
    last_interaction: {
      date: 2012-03-04T13:06:07.890Z,
      person_ids: [
        123,
        111,
      ],
    },
    next_event: {
      date: 2019-03-04T13:06:07.890Z,
      person_ids: [
        123,
        124,
        125,
      ],
    },
  },
  list_entries: [
    {
      created_at: 2015-12-11T10:26:56.537Z,
      creator_id: 38603,
      entity_id: 64779194,
      id: 389,
      list_id: 26,
    },
  ],
  name: "Affinity",
  opportunity_ids: [
    4,
    11,
  ],
  person_ids: [
    89734,
    117270,
    138123,
    274492,
    304848,
  ],
}
`;

snapshot[`organizations > can search for organizations 1`] = `
{
  next_page_token: "eyJwYXJhbXMiOnsidGVybSI6IiJ9LCJwYWdlX3NpemUiOjUsIm9mZnNldCI6MTB9",
  organizations: [
    {
      domain: "affinity.co",
      domains: [
        "affinity.co",
      ],
      global: false,
      id: 64779194,
      name: "Affinity",
    },
    {
      domain: "brandaffinity.net",
      domains: [
        "brandaffinity.net",
      ],
      global: true,
      id: 1513682,
      name: "Brand Affinity Technologies",
    },
  ],
}
`;

snapshot[`organizations > can create a new organization 1`] = `
{
  domain: "acme.co",
  domains: [
    "acme.co",
  ],
  global: false,
  id: 120611418,
  name: "Acme Corporation",
  person_ids: [
    38706,
  ],
}
`;

snapshot[`organizations > can update an organization 1`] = `
{
  domain: "acme.co",
  domains: [
    "acme.co",
  ],
  global: false,
  id: 120611418,
  name: "Acme Corp.",
  person_ids: [
    38706,
    89734,
  ],
}
`;

snapshot[`organizations > can delete an organization 1`] = `true`;

snapshot[`organizations > can get global organization fields 1`] = `
[
  {
    allows_multiple: false,
    dropdown_options: [],
    id: 662,
    name: "Potential Users",
    value_type: 3,
  },
  {
    allows_multiple: true,
    dropdown_options: [],
    id: 700,
    name: "Leads",
    value_type: 0,
  },
  {
    allows_multiple: false,
    dropdown_options: [],
    id: 2988,
    name: "Last Funding Date",
    value_type: 4,
  },
]
`;
